presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationBelowOuterJoin.java

```diff
@@ -134,8 +153,56 @@ private static boolean groupsOnAllOuterTableColumns(AggregationNode node, JoinNo
 
         private static PlanNode getFirstNonIdentityNode(PlanNode node)
         {
-            PlanNodeSearcher searcher = new PlanNodeSearcher(node).where(planNode -> !(planNode instanceof ProjectNode) || !((ProjectNode) planNode).isIdentity());
-            return searcher.findFirst().orElseThrow(() -> new IllegalStateException("no nodes below identity project"));
+            return searchFrom(node)
+                    .where(planNode -> !(planNode instanceof ProjectNode) || !((ProjectNode) planNode).isIdentity())
+                    .findFirst()
+                    .orElseThrow(() -> new IllegalStateException("no nodes below identity project"));
+        }
+
+        private static Optional<JoinNode> getFirstJoinNode(PlanNode node)
+        {
+            return searchFrom(node).skipOnlyWhen(planNode -> (planNode instanceof ProjectNode) && ((ProjectNode) planNode).isIdentity())
+                    .where(isInstanceOfAny(JoinNode.class))
+                    .findFirst();
+        }
+
+        private PlanNode addCoalesceIfNecessary(AggregationNode node, PlanNode result)
+        {
+            // The count functions return zero when the only input is null
+            List<Symbol> coalesceToZero = new ArrayList<>();
+            List<Symbol> coalesceToOne = new ArrayList<>();
+            List<String> zeroOnNullAggregations = ImmutableList.of("count_if", "approx_distinct");
```

Comment:
There is an another approach :) More hacky, but it does not require any function list to be maintained.

You can add yet another join :)

What would you tell about rewrittting plan to:
```
PROJECT (coalsce(agg_result, agg_result_on_empty_set))
   - CROSS JOIN 
      - LEFT JOIN (condition from correlation)
         - ...
         - AGGREGATION (some_agg(...) as agg_result, GROUP BY join condition keys)
           - ...
      - AGGREGATION(some_agg(...) as agg_result_on_empty_set, NO GROUP BY)
           - VALUES() -- an empty set
```

How do you like that?

Generally I would not botter that there is an additional CROSS JOIN and AGGREGATION at some point of time I think such plan should be constant folded ;). Even now it is so easy plan that it should not influence the performance.